index.js

```diff
@@ -52,322 +38,162 @@ class Client {
-      .then(configurePolicyForBucket)
-      .then(configureCorsForBucket)
-      .then(function() {
-        return this._uploadDirectory(this.clientPath);
+      })
+      .catch(error => {
+        return BbPromise.reject(new this.error(error));
       });
   }
 
-  _uploadDirectory(directoryPath) {
-    let _this = this,
-      readDirectory = _.partial(fs.readdir, directoryPath);
-
-    async.waterfall([
-      readDirectory,
-      function(files) {
-        files = _.map(files, function(file) {
-          return path.join(directoryPath, file);
-        });
-
-        async.each(files, function(path) {
-          fs.stat(
-            path,
-            _.bind(function(err, stats) {
-              return stats.isDirectory() ? _this._uploadDirectory(path) : _this._uploadFile(path);
-            }, _this)
-          );
-        });
-      }
-    ]);
-  }
-
-  _uploadFile(filePath) {
-    let _this = this,
-      urlRoot = regionToUrlRootMap(_this.region);
-
-    let fileKey = path.normalize(filePath).replace(_this.clientPath, '');
-    if (fileKey.substr(0, 1) === path.sep) {
-      fileKey = fileKey.replace(path.sep, '');
-    }
-
-    let distRoot = path.join(
-      _this.serverless.config.servicePath,
-      _this.serverless.service.custom.client.distributionFolder || path.join('client', 'dist')
-    );
-    distRoot += path.sep;
-
-    let baseHeaderKeys = [
-      'Cache-Control',
-      'Content-Disposition',
-      'Content-Encoding',
-      'Content-Language',
-      'Content-Type',
-      'Expires',
-      'Website-Redirect-Location'
-    ];
+  _processDeployment() {
+    this._validateConfig();
 
-    let ruleList = [];
-    let headers = _this.serverless.service.custom.client.objectHeaders;
-    if (headers) {
-      if (headers.ALL_OBJECTS) {
-        ruleList = ruleList.concat(headers.ALL_OBJECTS);
-      }
-      Object.keys(headers)
-        .filter(m => m.substr(-1, 1) === '/') // folders
-        .sort((a, b) => a.length > b.length) // sort by length ascending
-        .forEach(m => {
-          if (filePath.replace(distRoot, '').substr(0, m.length) === m) {
-            ruleList = ruleList.concat(headers[m]);
+    const region =
+      this.cliOptions.region ||
+      this.options.region ||
+      _.get(this.serverless, 'service.provider.region');
+    const distributionFolder = this.options.distributionFolder || path.join('client/dist');
+    const clientPath = path.join(this.serverless.config.servicePath, distributionFolder);
+    const bucketName = this.options.bucketName;
+    const headerSpec = this.options.objectHeaders;
+    const indexDoc = this.options.indexDocument || index.html;
+    const errorDoc = this.options.errorDocument || error.html;
+    const redirectAllRequestsTo = this.options.redirectAllRequestsTo || null;
+    const routingRules = this.options.routingRules || null;
+
+    this.serverless.cli.log(`Deploying client to bucket '${bucketName}' in region '${region}'...`);
+
+    const keepBucket =
```

Comment:
Can you explain this? I'm probably reading this wrong but it looks like keepBucket would only be true if they passed in a CLI argument to do one of the things below? Doesn't this, by default, evaluate to undefined?

Because I'm probably wrong and just confused here - throw in a comment to clarify.

Reply:
Ahh I think this is what you said before. This currently defaults to deleting the bucket?

Reply:
Right, we currently default to deleting the bucket but based on our other thread I think we want to go back to the previous functionality and always keep the bucket. So this code will be removed

Reply:
Awesome thanks!